anomalib/models/ganomaly/torch_model.py

```diff
@@ -0,0 +1,227 @@
+        for layer in range(extra_layers):
+            self.model.add_module(
+                f"extra-layers-{layer}-{n_features}-conv",
+                nn.Conv2d(n_features, n_features, kernel_size=3, stride=1, padding=1, bias=False),
+            )
+            self.model.add_module(f"extra-layers-{layer}-{n_features}-batchnorm", nn.BatchNorm2d(n_features))
+            self.model.add_module(f"extra-layers-{layer}-{n_features}-relu", nn.LeakyReLU(0.2, inplace=True))
+
+        # Create pyramid features to reach latent vector
+        input_size = input_size // 2
+        while input_size > 4:
+            in_features = n_features
+            out_features = n_features * 2
+            self.model.add_module(
+                f"pyramid-{in_features}-{out_features}-conv",
+                nn.Conv2d(in_features, out_features, kernel_size=4, stride=2, padding=1, bias=False),
+            )
+            self.model.add_module(f"pyramid-{out_features}-batchnorm", nn.BatchNorm2d(out_features))
+            self.model.add_module(f"pyramid-{out_features}-relu", nn.LeakyReLU(0.2, inplace=True))
+            n_features = out_features
+            input_size = input_size // 2
+
+        # Final conv
+        if add_final_conv_layer:
+            self.model.add_module(
+                f"final-{n_features}-{1}-conv",
+                nn.Conv2d(n_features, latent_vec_size, kernel_size=4, stride=1, padding=0, bias=False),
+            )
+
+    def forward(self, input_tensor: Tensor):
+        """Return latent vectors."""
+        return self.model(input_tensor)
+
+
+class Decoder(nn.Module):
+    """Decoder Network.
+
+    Args:
+        input_size (int): Size of input image
+        latent_vec_size (int): Size of latent vector z
+        num_input_channels (int): Number of input channels in the image
+        n_features (int): Number of features per convolution layer
+        extra_layers (int): Number of extra layers since the network uses only a single encoder layer by default.
+            Defaults to 0.
+    """
+
+    def __init__(
+        self, input_size: int, latent_vec_size: int, num_input_channels: int, n_features: int, extra_layers: int = 0
+    ):
+        super().__init__()
+        assert input_size % 16 == 0, "Input size should be a multiple of 16"  # Why again?
+
+        self.model = nn.Sequential()
+
+        # Calculate input channel size to recreate inverse pyramid
+        exp_factor = int(math.log(input_size // 4, 2)) - 1
+        n_input_features = n_features * (2 ** exp_factor)
+
+        # CNN layer for latent vector input
+        self.model.add_module(
+            f"initial-{latent_vec_size}-{n_input_features}-convt",
+            nn.ConvTranspose2d(latent_vec_size, n_input_features, kernel_size=4, stride=1, padding=0, bias=False),
+        )
+        self.model.add_module(f"initial-{n_input_features}-batchnorm", nn.BatchNorm2d(n_input_features))
+        self.model.add_module(f"initial-{n_input_features}-relu", nn.ReLU(True))
```

Comment:
Same as in `Encoder`. I think it would be good to split `self.model` up into smaller chunks.